apps/glusterfs/operations.go

```diff
@@ -980,6 +1010,10 @@ func RunOperation(o Operation,
 		return err
 	}
 	if err := o.Exec(executor); err != nil {
+		if _, ok := err.(OperationRetryError); ok && o.MaxRetries() > 0 {
```

Comment:
Somehow, the way it's done feels a little bit backward and redundant. Can't we make it more natural by wrapping it the other way to remove code duplication, like so (simplified):

```
func RunOperation(o, e) (err) {
    max := o.MaxRetries()
    for i := 0; i < max + 1; i++ {
        if e := o.Build(); e != nil {
            return e
        }

        err = o.Exec()
        if err == nil {
            return o.Finalize()
        }

        if re := o.Rollback(); re != nil {
            return re
        }

        if _, ok := err.(OperationRetryError); !ok {
            break
        }
    }

    // too many retries or not OperationRetryError
    return err
}
```

Of course for the async form this would need to be adapted...

Reply:
This is a very good point and I agree that the current loop is a bit funny. However, it might be better to take this PR and clean up these loops in a followup. That way we get some testing around the overall feature for volume create retries and we validate the overall concept before cleaning up this part.

Reply:
Agreed!